### PR TITLE
Some improvements to community.general.telegram

### PR DESCRIPTION
**Pull request by LOMS**
##### SUMMARY
Telegram API changed drastically since 2016. I made some changes to cover latest and future changes in API.
Backward compatibility preserved.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
community.general.telegram

##### ADDITIONAL INFORMATION
see https://core.telegram.org/bots/api#sendmessage
Personally, I missed following new options for "SendMessage":
disable_web_page_preview
disable_notification
parse_mode == MarkdownV2 (wasn't an option)

I also made "api_method" as argument, so you may use other api methods if needed.

So, before we could do only this:
```   
- name: Send a message to chat in playbook (old style)
  community.general.telegram:
    token: '9999999:XXXXXXXXXXXXXXXXXXXXXXX'
    chat_id: 000000
    msg: Ansible task finished
```
now you can:
```
- name: Send notify to Telegram
  community.general.telegram:
    token: '9999999:XXXXXXXXXXXXXXXXXXXXXXX'
    api_args:
      chat_id: 000000
      parse_mode: "markdown"
      text: "Your precious application has been deployed: https://example.com"
      disable_web_page_preview: True
      disable_notification: True
```
Also, you may call other API methods like this:
```

- name: Forward message to someone
  community.general.telegram:
    token: '9999999:XXXXXXXXXXXXXXXXXXXXXXX'
    api_method: forwardMessage
    api_args:
      chat_id: 000000
      from_chat_id: 111111
      disable_notification: True
      message_id: '{{ saved_msg_id }}'
```

P.S.: new contributor, probably I'm missing something in my own PR, like documentation or tests (still have no idea how I test this). I'm open to suggestions and criticism.

### CODE REVIEW

File: plugins/modules/notification/telegram.py [hunk 1]
```diff
@@ -11,44 +11,76 @@
...
+author:
+ - "Artem Feofanov (@tyouxa)"
+ - "Nikolai Lomov (@lomserman)"
 
 short_description: module for sending notifications via telegram
 
 description:
-    - Send notifications via telegram bot, to a verified group or user
+    - Send notifications via telegram bot, to a verified group or user.
+    - Also, you may try to use any other telegram bot api method, if you specify "api_method" argument.
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:22]:
```suggestion
    - Also, the user may try to use any other telegram bot API method, if you specify I(api_method) argument.
```

File: plugins/modules/notification/telegram.py [hunk 2]
```diff
@@ -11,44 +11,76 @@
...
+      This isn't tested though.
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:23]:
```suggestion
      This is not tested though.
```

File: plugins/modules/notification/telegram.py [hunk 3]
```diff
@@ -11,44 +11,76 @@
...
 notes:
     - You will require a telegram account and create telegram bot to use this module.
+    - Deprecated arguments left for backward compatibility, when used - will replace corresponding
+      variables in api_args
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:27]:
```suggestion
      variables in I(api_args).
```

File: plugins/modules/notification/telegram.py [hunk 4]
```diff
@@ -11,44 +11,76 @@
...
-  msg:
+  token:
     type: str
     description:
-      - What message you wish to send.
+      - Token identifying your telegram bot.
     required: true
+  api_method:
+    type: str
+    description: BOt API method. For reference, see https://core.telegram.org/bots/api
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:36]:
```suggestion
    description:
    - Bot API method.
    - For reference, see U(https://core.telegram.org/bots/api).
```

File: plugins/modules/notification/telegram.py [hunk 5]
```diff
@@ -11,44 +11,76 @@
...
+    default: SendMessage
+  api_args:
+    type: dict
+    description:
+      - Any parameters for method. I.e., for SendMessage, see https://core.telegram.org/bots/api#sendmessage
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:41]:
```suggestion
      - Any parameters for the method. For example, for C(SendMessage), see U(https://core.telegram.org/bots/api#sendmessage).
```

File: plugins/modules/notification/telegram.py [hunk 6]
```diff
@@ -11,44 +11,76 @@
...
+  msg:
+    type: str
+    description:
+      - (Deprecated) What message you wish to send.
```
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:45]: @felixfontein Could you please comment on the deprecation cycle guidelines?

File: plugins/modules/notification/telegram.py (hunk 1 above)
Comment (Akasurde) [inline plugins/modules/notification/telegram.py:22]: Add a change log entry for this change like - https://github.com/ansible-collections/community.general/tree/main/changelogs/fragments

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: plugins/modules/notification/telegram.py [hunk 7]
```diff
@@ -11,44 +11,76 @@
...
   msg_format:
     type: str
     description:
-      - Message format. Formatting options `markdown` and `html` described in
+      - (Deprecated) Message format. Formatting options `markdown`, `MarkdownV2` and `html` described in
```
Comment (Andersson007) [inline plugins/modules/notification/telegram.py:49]:
```suggestion
      - (Deprecated) Message format. Formatting options C(markdown), C(MarkdownV2), and C(html) described in
```

File: plugins/modules/notification/telegram.py [hunk 8]
```diff
@@ -11,44 +11,76 @@
...
-    description:
-      - Token identifying your telegram bot.
-    required: true
+    choices: [ "plain", "markdown", "MarkdownV2", "html" ]
   chat_id:
     type: str
     description:
-      - Telegram group or user chat_id
-    required: true
+      - (Deprecated) Telegram group or user chat_id
```
Comment (Andersson007) [inline plugins/modules/notification/telegram.py:57]:
```suggestion
      - (Deprecated) Telegram group or user chat_id.
```

File: plugins/modules/notification/telegram.py [hunk 9]
```diff
@@ -11,44 +11,76 @@
...
+    type: str
+    description:
+      - (Deprecated) What message you wish to send.
   msg_format:
     type: str
     description:
-      - Message format. Formatting options `markdown` and `html` described in
+      - (Deprecated) Message format. Formatting options `markdown`, `MarkdownV2` and `html` described in
         Telegram API docs (https://core.telegram.org/bots/api#formatting-options).
         If option `plain` set, message will not be formatted.
```
Comment (Andersson007) [inline plugins/modules/notification/telegram.py:51]:
```suggestion
        If option C(plain) set, message will not be formatted.
```

File: plugins/modules/notification/telegram.py (hunk 1 above)
Comment (LOMS) [reply plugins/modules/notification/telegram.py:22]: @Akasurde fragment added. Thx for review btw.

File: changelogs/fragments/telegram-api-update.yml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+  - community.general.telegram - now can call any methods in Telegram bot API.
```
Comment (felixfontein) [inline changelogs/fragments/telegram-api-update.yml:2]:
```suggestion
  - telegram - now can call any methods in Telegram bot API.
```

File: changelogs/fragments/telegram-api-update.yml [hunk 11]
```diff
@@ -0,0 +1,5 @@
...
+    Previously this module was hardcoded to use "SendMessage" only.
+  - With previous change, usage of "SendMessage" API method was also librated,
+    and now you can specify any arguments you need, for example, "disable_notificaton".
```
Comment (felixfontein) [inline changelogs/fragments/telegram-api-update.yml:5]:
```suggestion
    and now you can specify any arguments you need, for example, "disable_notificaton" (https://github.com/ansible-collections/community.general/pull/1642).
```

File: changelogs/fragments/telegram-api-update.yml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+minor_changes:
+  - community.general.telegram - now can call any methods in Telegram bot API.
+    Previously this module was hardcoded to use "SendMessage" only.
+  - With previous change, usage of "SendMessage" API method was also librated,
```
Comment (felixfontein) [inline changelogs/fragments/telegram-api-update.yml:4]:
```suggestion
    Usage of "SendMessage" API method was also librated,
```

File: plugins/modules/notification/telegram.py [hunk 13]
```diff
@@ -11,44 +11,77 @@
...
     description:
-      - What message you wish to send.
+      - Token identifying your telegram bot.
     required: true
+  api_method:
+    type: str
+    description:
+    - Bot API method.
+    - For reference, see U(https://core.telegram.org/bots/api).
+    default: SendMessage
```
Comment (felixfontein) [inline plugins/modules/notification/telegram.py:38]:
```suggestion
    default: SendMessage
    version_added: 2.0.0
```

File: plugins/modules/notification/telegram.py [hunk 14]
```diff
@@ -11,44 +11,77 @@
...
+  api_args:
+    type: dict
+    description:
+      - Any parameters for the method. For example, for C(SendMessage), see U(https://core.telegram.org/bots/api#sendmessage).
```
Comment (felixfontein) [inline plugins/modules/notification/telegram.py:42]:
```suggestion
      - Any parameters for the method. For example, for C(SendMessage), see U(https://core.telegram.org/bots/api#sendmessage).
    version_added: 2.0.0
```

File: plugins/modules/notification/telegram.py [hunk 15]
```diff
@@ -11,44 +11,77 @@
...
+  community.general.telegram:
+    token: '9999999:XXXXXXXXXXXXXXXXXXXXXXX'
+    api_method: forwardMessage
+    api_args:
+      chat_id: 000000
+      from_chat_id: 111111
+      disable_notification: True
+      message_id: '{{ saved_msg_id }}'
+
+- name: Send a message to chat in playbook (old style)
```
Comment (felixfontein) [inline plugins/modules/notification/telegram.py:84]:
```suggestion
- name: Send a message to chat in playbook (deprecated old style)
```

File: plugins/modules/notification/telegram.py [hunk 16]
```diff
@@ -11,44 +11,77 @@
...
 short_description: module for sending notifications via telegram
 
 description:
-    - Send notifications via telegram bot, to a verified group or user
+    - Send notifications via telegram bot, to a verified group or user.
+    - Also, the user may try to use any other telegram bot API method, if you specify I(api_method) argument.
 notes:
     - You will require a telegram account and create telegram bot to use this module.
+    - Deprecated arguments left for backward compatibility, when used - will replace corresponding
+      variables in I(api_args).
```
Comment (felixfontein) [inline plugins/modules/notification/telegram.py:26]:
```suggestion
    - The options I(msg), I(msg_format) and I(chat_id) have been deprecated and will be removed in community.general 4.0.0. Use the corresponding
      variables in I(api_args) instead. See the examples for how that works.
```

File: plugins/modules/notification/telegram.py [hunk 17]
```diff
@@ -72,42 +105,58 @@
...
-            chat_id=dict(type='str', required=True, no_log=True),
-            msg_format=dict(type='str', required=False, default='plain',
-                            choices=['plain', 'markdown', 'html']),
-            msg=dict(type='str', required=True)),
+            api_args=dict(type='dict'),
+            api_method=dict(type="str", default="SendMessage"),
+            # backwards compatibility
+            chat_id=dict(type='str', no_log=True),
+            msg=dict(type='str'),
+            msg_format=dict(type='str', choices=['plain', 'markdown', 'html', 'MarkdownV2'], default='plain'),
```
Comment (felixfontein) [inline plugins/modules/notification/telegram.py:122]:
```suggestion
            chat_id=dict(type='str', no_log=True, removed_in_version='4.0.0', removed_from_collection='community.general'),
            msg=dict(type='str', removed_in_version='4.0.0', removed_from_collection='community.general'),
            msg_format=dict(type='str', choices=['plain', 'markdown', 'html', 'MarkdownV2'], default='plain', removed_in_version='4.0.0', removed_from_collection='community.general'),
```

**Review by felixfontein — COMMENTED** [review]
I'm not sure whether I like this PR. It mixes a lot of different information into a completely not validated dict (`api_args`). Some of the data that goes in there now used to be `no_log=True`, but most not. To avoid sensitive information to be leaked, we would need to set `no_log=True` for `api_args`, but that makes the module harder to debug.

Maybe the module should be kept more as it was before, and instead an additional `telegram_api` module should be added for advanced usage?

**Comment by LOMS** [comment]
> I'm not sure whether I like this PR. It mixes a lot of different information into a completely not validated dict (`api_args`). Some of the data that goes in there now used to be `no_log=True`, but most not. To avoid sensitive information to be leaked, we would need to set `no_log=True` for `api_args`, but that makes the module harder to debug.
> 
> Maybe the module should be kept more as it was before, and instead an additional `telegram_api` module should be added for advanced usage?

I do believe the only sensitive information is bot token. Chat Id isn't really 'sensitive', because private chat's ID gives you nothing, unless you're invited to conversation. And public chat's ID are public. Also, if you keep using old arguments, chat id is ******* in log despite all my changes (I never go deep into that, just considered it some weird ansible magic and kept going further)

File: plugins/modules/notification/telegram.py (hunk 6 above)
Comment (LOMS) [reply plugins/modules/notification/telegram.py:49]: Resolved somewhere below.

**Comment by LOMS** [comment]
@Akasurde it says you "requested changes", but I can't see any requested changes left. Am I missing something?

**Review by Akasurde — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@Andersson007 if you're happy, feel free to merge this so it can get included into 2.0.0.

**Comment by Andersson007** [comment]
@LOMS thanks for this! I'll add you as a maintainer to .github.com/BOTMETA.yml so that you'll get notifications if there are any issues related to the module and your `shipit` will be counted by the bot for fixes / new features automerge, do you agree?

@felixfontein @Akasurde thanks for reviewing!

**Comment by LOMS** [comment]
@Andersson007 ...I added myself to BOTMETA a while ago, so I agree =D

**Comment by Andersson007** [comment]
@LOMS ah, ok, i see:D thanks (@felixfontein we should ask users to do the same themselves in every PR:))

**Pull request merged**